ql/src/java/org/apache/hadoop/hive/ql/parse/GenTezUtils.java

```diff
@@ -182,6 +187,41 @@ public static ReduceWork createReduceWork(
     return reduceWork;
   }
 
+  /**
+   * Checks if there is a Bucket Map Join (BMJ) following a hierarchy:
+   * ReduceSinkOperator (RS) -> ReduceSinkOperator (RS) -> MapJoinOperator (MJ).
+   * Ensures at most **two** RSOs before MJ.
```

Comment:
I guess it happens only when there are precisely two RSOs. Given that Map 3 -> Reducer 4 -> Reducer 5 -> Map 1(BMJ), `createReduceWork` is invoked twice.
When we create Reducer 4, the passed RSO belongs to Map 3. The process bails out once we find the third RSO in Reducer 5.
When we create Reducer 5, we find two RSOs in Reduer 4 and Reducer 5, and the BucketMapJoin.


A vertex with BucketMapJoin is a mapper. So, basically, RS -> MJ doesn't happen.

Reply:
Thanks